### PR TITLE
feat: improve map(), struct(), array()

### PR DESCRIPTION
**Pull request by NickCrews**
fixes https://github.com/ibis-project/ibis/issues/8289

This does a lot of changes. It was hard for me to separate them out as I implemented them. But now that it's all hashed out, I can try to split this up into separate commits if you want. But that might be sorta hard in
some cases.

## Changes:

This is adding support for passing in None to all these constructors.
These use the new `ibis.null(<type>)` API to return `op.Literal(None, <type>)`s

Make these constructors idempotent: you can pass in existing Expressions into array(), etc.
The type argument for all of these now always has an effect, not just when passing in python literals. So basically it acts like a cast.

A big structural change is that now ops.Array has an optional
attribute "dtype", so if you pass in a 0-length sequence
of values the op still knows what dtype it is.

Several of the backends were always broken here, they just weren't getting caught. I marked them as broken, we can fix them in a followup.

You can test this locally with eg
`pytest -m <backend> -k factory ibis/backends/tests/test_array.py  ibis/backends/tests/test_map.py ibis/backends/tests/test_struct.py`

Also, fix executing Literal(None) on pandas and polars, 0-length arrays on polars

## Next steps

A possible nice follow-up would be to completely remove supporting structs, arrays, and maps inside ops.Literal.
Currently, when someone calls `ibis.array([1,2])`, that now results in ops.Array, but if they do `ibis.literal([1,2])`, it is represented as ops.Literal. It would be nice if inside ibis.literal() there was some logic like

```python
...
if isinstance(val, Mapping):
    return ibis.map(val)
if is_iterable(val):
    return ibis.array(val)
...
```

and then all the backends wouldn't have to even worry about supporting these nested types in `visit_Literal()`

### CODE REVIEW

**Review by jcrist — COMMENTED** [review]
Makes sense to me. Can you add a few tests in the appropriate `ibis/tests/expr/test_{dtype}.py` files?

Also curious - what was the use case that led to you wanting this change?

**Comment by NickCrews** [comment]
in my first attempt at https://github.com/ibis-project/ibis/pull/8649, [I went about it the wrong way](https://github.com/ibis-project/ibis/commit/6925c908122174cd3e6690b50aa2777b6b6463d1#diff-d50c70a1d9c6f811ab5dba8574dbb12647c81ced87d539b19f5517ec037eca9eR495-R496), but I needed to ensure something was an Array, but the incoming thing might already have been an array.

File: ibis/expr/types/maps.py [hunk 1]
```diff
@@ -435,8 +436,10 @@ def __getitem__(self, key: ir.Value) -> ir.Column:
 @public
 @deferrable
 def map(
-    keys: Iterable[Any] | Mapping[Any, Any] | ArrayColumn,
-    values: Iterable[Any] | ArrayColumn | None = None,
+    keys: Iterable[Any] | Mapping[Any, Any] | ir.ArrayValue | MapValue | None,
+    values: Iterable[Any] | ir.ArrayValue | None = None,
+    *,
+    type: str | dt.DataType | None = None,
```
Comment (NickCrews) [inline ibis/expr/types/maps.py:442]: For this factory, `type` is kwarg-only. Should the other factories also be made kwarg-only to match?

File: ibis/expr/types/arrays.py [hunk 2]
```diff
@@ -1120,15 +1128,32 @@ def array(values: Iterable[V]) -> ArrayValue:
...
+    if values is None:
+        if type is None:
+            raise TypeError("type must be specified when values is None")
+        return ibis.literal(None, type=type)
+
+    if isinstance(values, ir.ArrayValue):
+        result = values
+    else:
+        values = tuple(values)
+        if len(values) == 0:
```
Comment (cpcloud) [inline ibis/expr/types/arrays.py:1151]:
```suggestion
        if not values:
```

File: ibis/backends/tests/test_struct.py [hunk 3]
```diff
@@ -23,6 +24,38 @@
 ]
 
 
+@pytest.mark.notimpl(["postgres"])
+# @pytest.mark.broken(["pandas", "dask"], reason="casting is broken")
```
Comment (cpcloud) [inline ibis/backends/tests/test_struct.py:28]:
```suggestion
```

File: ibis/backends/tests/test_array.py [hunk 4]
```diff
@@ -70,6 +70,40 @@
...
+    assert con.execute(typed2) == [1.0, 2.0, 3.0]
+
+
+@pytest.mark.notimpl("postgres", raises=PsycoPg2IndeterminateDatatype)
+def test_array_factory_empty(con):
+    with pytest.raises(TypeError):
+        ibis.array([])
+
+    empty_typed = ibis.array([], type="array<float64>")
+    assert str(empty_typed.type()) == "array<float64>"
```
Comment (cpcloud) [inline ibis/backends/tests/test_array.py:90]:
```suggestion
```

If you want to verify the type, use a datatype instance, not its string representation.

File: ibis/backends/tests/test_map.py [hunk 5]
```diff
@@ -26,6 +30,48 @@
...
+    ["risingwave"],
+    raises=PsycoPg2InternalError,
+    reason="function hstore(character varying[], character varying[]) does not exist",
+)
+@pytest.mark.notyet("postgres", reason="only support maps of string -> string")
+def test_map_factory_empty(con):
+    with pytest.raises(TypeError):
+        ibis.map({})
+    empty_typed = ibis.map({}, type="map<string, string>")
+    assert str(empty_typed.type()) == "map<string, string>"
```
Comment (cpcloud) [inline ibis/backends/tests/test_map.py:60]:
```suggestion
```

Construct the type using a Python object, not a string.

File: ibis/backends/tests/test_map.py [hunk 6]
```diff
@@ -26,6 +30,48 @@
...
+
+
+@pytest.mark.notyet(
+    "clickhouse", raises=ClickHouseDatabaseError, reason="nested types can't be NULL"
+)
+def test_map_factory_null(con):
+    with pytest.raises(TypeError):
+        ibis.map(None)
+    null_typed = ibis.map(None, type="map<string, string>")
+    assert str(null_typed.type()) == "map<string, string>"
```
Comment (cpcloud) [inline ibis/backends/tests/test_map.py:71]:
```suggestion
```

File: ibis/backends/tests/test_struct.py [hunk 7]
```diff
@@ -23,6 +24,38 @@
...
+
+@pytest.mark.broken("polars", raises=AttributeError)
+@pytest.mark.notyet(
+    "clickhouse", raises=ClickHouseDatabaseError, reason="nested types can't be NULL"
+)
+def test_struct_factory_null(con):
+    with pytest.raises(TypeError):
+        ibis.struct(None)
+    none_typed = ibis.struct(None, type="struct<a: float64, b: float>")
+    assert str(none_typed.type()) == "struct<a: float64, b: float64>"
```
Comment (cpcloud) [inline ibis/backends/tests/test_struct.py:55]:
```suggestion
```

File: ibis/expr/types/arrays.py [hunk 8]
```diff
@@ -5,14 +5,16 @@
...
 import ibis.expr.operations as ops
+import ibis.expr.types as ir
 from ibis.common.deferred import Deferred, deferrable
 from ibis.expr.types.generic import Column, Scalar, Value
 
 if TYPE_CHECKING:
     from collections.abc import Iterable
 
-    import ibis.expr.types as ir
+    from ibis.expr.types import dt
```
Comment (cpcloud) [inline ibis/expr/types/arrays.py:17]:
```suggestion
    import ibis.expr.datatypes as dt
```

Definitely do not import internal module aliases from other modules.

File: ibis/expr/types/maps.py [hunk 9]
```diff
@@ -4,15 +4,16 @@
...
+import ibis.expr.types as ir
 from ibis.common.deferred import deferrable
 from ibis.expr.types.generic import Column, Scalar, Value
 
 if TYPE_CHECKING:
     from collections.abc import Iterable, Mapping
 
-    import ibis.expr.types as ir
-    from ibis.expr.types.arrays import ArrayColumn
+    from ibis.expr import datatypes as dt
```
Comment (cpcloud) [inline ibis/expr/types/maps.py:16]: Please follow the same import style as the other Ibis modules:

```suggestion
    import ibis.expr.datatype as dt
```

File: ibis/expr/types/maps.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/types/maps.py:442]: Yes, make them consistent :)

File: ibis/expr/types/structs.py [hunk 10]
```diff
@@ -62,26 +61,38 @@ def struct(
...
+        if type is None:
+            raise TypeError("Must specify type if value is None")
+        return literal(None, type=type)
+
+    if isinstance(value, StructValue):
+        result = value
     else:
-        return literal(collections.OrderedDict(fields), type=type)
+        fields = dict(value)
+        if any(isinstance(value, Value) for value in fields.values()):
```
Comment (cpcloud) [inline ibis/expr/types/structs.py:86]: We should really get rid of this branching here and not have this odd split between literal structs and "non-literal" structs. There's no real reason to differentiate AFAICT.

That can be done in a follow-up.

File: ibis/formats/pandas.py [hunk 11]
```diff
@@ -363,6 +363,11 @@ def convert(values):
             if values is None:
                 return values
 
+            try:
+                values = iter(values)
+            except TypeError:
+                # iteration over 0-d array, happens when the element is NULL
+                return None
```
Comment (cpcloud) [inline ibis/formats/pandas.py:370]: Did you add a test case for this code path?

File: ibis/backends/tests/test_array.py [hunk 12]
```diff
@@ -70,6 +70,40 @@
...
 
+def test_array_factory(con):
+    a = ibis.array([1, 2, 3])
+    assert con.execute(a) == [1, 2, 3]
+    a2 = ibis.array(a)
+    assert con.execute(a2) == [1, 2, 3]
+    typed = ibis.array([1, 2, 3], type="array<float64>")
+    assert con.execute(typed) == [1.0, 2.0, 3.0]
+    typed2 = ibis.array(a, type="array<float64>")
+    assert con.execute(typed2) == [1.0, 2.0, 3.0]
```
Comment (cpcloud) [inline ibis/backends/tests/test_array.py:81]: Can you group each pair of lines with a newline between them? It's hard to read at the moment.

**Review by cpcloud — CHANGES_REQUESTED** [review]

File: ibis/backends/pandas/convert.py [hunk 13]
```diff
@@ -83,3 +83,11 @@ def convert_String(cls, s, dtype, pandas_type):
         # TODO(kszucs): should switch to the new pandas string type and convert
         # object columns using s.convert_dtypes() method
         return s.map(str, na_action="ignore").astype(object)
+
+    @classmethod
+    def convert_Struct(cls, s, dtype, pandas_type):
+        # dt.normalize returns a frozendict, which doesn't play well
+        # with the pandas executor
```
Comment (jcrist) [inline ibis/backends/pandas/convert.py:90]: It's just the formatter when pandas repr's a series containing a `FrozenDict`, the actual execution is fine. See #8687.

File: ibis/backends/pandas/convert.py [hunk 14]
```diff
@@ -83,3 +83,11 @@ def convert_String(cls, s, dtype, pandas_type):
...
+        # https://github.com/ibis-project/ibis/issues/8687
+        func = lambda x: x if x is pd.NA else dict(dt.normalize(dtype, x))
```
Comment (jcrist) [inline ibis/backends/pandas/convert.py:92]: I _think_ you'll need to recurse here for nested structs.

File: ibis/backends/pandas/convert.py (hunk 13 above)
Comment (NickCrews) [reply ibis/backends/pandas/convert.py:90]: It actually isn't, if you leave out this method the test fails. See the linked issue again, I added some clarification

File: ibis/backends/pandas/convert.py (hunk 14 above)
Comment (NickCrews) [reply ibis/backends/pandas/convert.py:92]: dt.normalize will recurse, but that's not quite enough, that would lead to dict(frozendict(frozendict(..))). I assume we want dict all the way down. This seems hard to do.

better if we can address the linked bug, then just frozendict s all the way down and we don't need this workaround

File: ibis/backends/tests/test_struct.py (hunk 3 above)
Comment (NickCrews) [reply ibis/backends/tests/test_struct.py:28]: here is the test case that exercises the change in formats/pandas.py.

File: ibis/formats/pandas.py (hunk 11 above)
Comment (NickCrews) [reply ibis/formats/pandas.py:370]: Yes, test_struct_factory fails for pandas and dask without this change. That is why there was that stray forgotten commented out `xfail(["pandas", "dask"])` on that test.

File: ibis/expr/types/maps.py (hunk 9 above)
Comment (NickCrews) [reply ibis/expr/types/maps.py:16]: oops, sorry this was blind acceptance of either vscode intellisense or copilor suggestsions.

File: ibis/expr/types/arrays.py (hunk 8 above)
Comment (NickCrews) [reply ibis/expr/types/arrays.py:17]: oops, must have been a blind vscode/copilot acceptance.

File: ibis/backends/tests/test_array.py (hunk 12 above)
Comment (NickCrews) [reply ibis/backends/tests/test_array.py:84]: done.

File: ibis/backends/tests/test_array.py (hunk 4 above)
Comment (NickCrews) [reply ibis/backends/tests/test_array.py]: done

File: ibis/backends/pandas/convert.py (hunk 14 above)
Comment (NickCrews) [reply ibis/backends/pandas/convert.py]: This has been fixed in [#Frozen](https://github.com/ibis-project/ibis/pull/8693) so I removed this workaround

File: ibis/backends/pandas/convert.py (hunk 13 above)
Comment (NickCrews) [reply ibis/backends/pandas/convert.py]: resolved in https://github.com/ibis-project/ibis/pull/8693, so this is no longer relevant.

File: ibis/formats/pandas.py (hunk 11 above)
Comment (NickCrews) [reply ibis/formats/pandas.py]: this is removed now

File: ibis/expr/types/structs.py (hunk 10 above)
Comment (NickCrews) [reply ibis/expr/types/structs.py]: OK, I ended up doing this myself in this commit. The branching was causing me lots of problems too.

File: ibis/backends/tests/test_struct.py (hunk 3 above)
Comment (NickCrews) [reply ibis/backends/tests/test_struct.py]: no longer relevant.

**Comment by NickCrews** [comment]
OK I just pushed an absolute doozy of a commit in https://github.com/ibis-project/ibis/pull/8666/commits/530fad0e29d37b52880b2767f527d3abc8790fdb. Curious how many backends that fails on. It required really overhauling the internal representation of ops.Array and ops.Struct to be able to distinguish between
- the overall value is NULL
- the contained values is NULL
Take a look at that commit description.

**Comment by ncclementi** [comment]
Hi @NickCrews, checking in here. What's the status on this one? It looks like there are a lot of conflicts and red on CI 😅 .

**Comment by NickCrews** [comment]
Yeeesssss this needs a bit of love :) Once my other pr with .cases lands I figured I would pick this up again.

File: ibis/backends/dask/helpers.py [hunk 15]
```diff
@@ -47,7 +47,7 @@ def asseries(cls, value, like=None):
         elif isinstance(value, pd.Series):
             return dd.from_pandas(value, npartitions=1)
         elif like is not None:
-            if isinstance(value, (tuple, list, dict)):
+            if isinstance(value, (tuple, list, dict, np.ndarray)):
```
Comment (NickCrews) [inline ibis/backends/dask/helpers.py:50]: ops.Array executes to a np.ndarray (both before this change, and after). We never ran into this scenario in tests before, so this tweak here is adjusting an existing bug.

**Comment by NickCrews** [comment]
@cpcloud @jcrist this is ready for review whenever you get the chance. This is a bit of a doozy that might want to get split up into multiple commits/PRs. But I think it would be good to get a check that the overall direction is right before I do that? The summary in the top of this PR is up to date for a summary. Maybe read that and then it would be probably best to look at the tests, and then the factory methods in types/arrays.py, types/maps.py, and types/structs.py

**Comment by cpcloud** [comment]
@NickCrews 

> Also, fix a typing bug: map() can accept ArrayValues, not just ArrayColumns.

> Also, fix executing Literal(None) on pandas and polars, 0-length arrays on polars

> Also, fixing converting dtypes on clickhouse, Structs should be converted to nonnullable dtypes.

> Also, implement ops.StructColumn on pandas and dask

Please split these up into separate PRs.

Is 

> A big structural change is that now ops.Array has an optional
attribute "dtype", so if you pass in a 0-length sequence
of values the op still knows what dtype it is.

required for this PR?

Can't we handle this in separate PR? It seems like it's just to handle the empty array case.

**Comment by NickCrews** [comment]
> Please split these up into separate PRs.

Yeah will do :)

> It seems like it's just to handle the empty array case.

It is. This is because currently empty arrays are represented by ops.Literals. This leads to needing to compile arrays inside visit_Literal. After this PR, ops.Array can handle empty arrays (ops.Struct and ops.Map already can I think), so we are then free to move towards removing all nested types from ops.Literal, and visit_Literal will only need to deal with non-nested values. What do you think, is this worth the churn?

**Comment by cpcloud** [comment]
> visit_Literal will only need to deal with non-nested values. What do you think, is this worth the churn?

I think this is worth exploring for sure. I'm not 100% sure it will be a net improvement without implementing it though :) It sounds like a nice improvement if it works out though.

File: ibis/backends/pandas/executor.py [hunk 16]
```diff
@@ -49,12 +49,14 @@ def visit(cls, op: ops.Node, **kwargs):
...
     @classmethod
     def visit(cls, op: ops.Literal, value, dtype):
+        if value is None:
+            return None
         if dtype.is_interval():
-            value = pd.Timedelta(value, dtype.unit.short)
-        elif dtype.is_array():
-            value = np.array(value)
-        elif dtype.is_date():
-            value = pd.Timestamp(value, tz="UTC").tz_localize(None)
```
Comment (cpcloud) [inline ibis/backends/pandas/executor.py:57]: Can we keep the previous style? The branches are intentionally written this way.

File: ibis/backends/polars/compiler.py [hunk 17]
```diff
@@ -86,10 +86,14 @@ def _make_duration(value, dtype):
 def literal(op, **_):
     value = op.value
     dtype = op.dtype
+    if dtype.is_interval():
+        return _make_duration(value, dtype)
 
-    if dtype.is_array():
```
Comment (cpcloud) [inline ibis/backends/polars/compiler.py:90]: Can you avoid moving the code around here and **only** handle the `None` case? It's harder to review when these kinds of changes are mixed together.

File: ibis/backends/polars/compiler.py [hunk 18]
```diff
@@ -980,9 +981,11 @@ def array_concat(op, **kw):
 
 
 @translate.register(ops.Array)
-def array_column(op, **kw):
-    cols = [translate(col, **kw) for col in op.exprs]
-    return pl.concat_list(cols)
+def array_literal(op, **kw):
+    if len(op.exprs) > 0:
```
Comment (cpcloud) [inline ibis/backends/polars/compiler.py:985]:
```suggestion
    if op.exprs:
```

File: ibis/backends/sql/compiler.py [hunk 19]
```diff
@@ -970,8 +970,11 @@ def visit_InSubquery(self, op, *, rel, needle):
             query = sg.select(STAR).from_(query)
         return needle.isin(query=query)
 
-    def visit_Array(self, op, *, exprs):
-        return self.f.array(*exprs)
+    def visit_Array(self, op, *, exprs, dtype):
+        result = self.f.array(*exprs)
+        if len(exprs) == 0:
```
Comment (cpcloud) [inline ibis/backends/sql/compiler.py:975]:
```suggestion
        if not exprs:
```

File: ibis/backends/sql/datatypes.py [hunk 20]
```diff
@@ -1007,8 +1007,10 @@ class ClickHouseType(SqlglotType):
     def from_ibis(cls, dtype: dt.DataType) -> sge.DataType:
         """Convert a sqlglot type to an ibis type."""
         typ = super().from_ibis(dtype)
-        if dtype.nullable and not (dtype.is_map() or dtype.is_array()):
-            # map cannot be nullable in clickhouse
+        # nested types cannot be nullable in clickhouse
+        if dtype.nullable and not (
+            dtype.is_map() or dtype.is_array() or dtype.is_struct()
+        ):
```
Comment (cpcloud) [inline ibis/backends/sql/datatypes.py:1013]: Can you break this out into a separate PR?

File: ibis/backends/tests/test_array.py [hunk 21]
```diff
@@ -72,6 +73,75 @@
...
+    assert a.type() == dt.Array(value_type=dt.Int8)
+    assert con.execute(a) == [1, 2, 3]
+
+    a2 = ibis.array(a)
+    assert a.type() == dt.Array(value_type=dt.Int8)
+    assert con.execute(a2) == [1, 2, 3]
+
+
+def test_array_factory_typed(con):
+    typed = ibis.array([1, 2, 3], type="array<string>")
```
Comment (cpcloud) [inline ibis/backends/tests/test_array.py:87]: I don't really like the implicit conversion here. Is it necessary?

File: ibis/expr/operations/structs.py [hunk 22]
```diff
@@ -34,7 +34,9 @@ class StructColumn(Value):
 
     shape = rlz.shape_like("values")
 
-    def __init__(self, names, values):
+    def __init__(self, names: VarTuple[str], values: VarTuple[Value]):
+        if len(names) == 0:
+            raise ValidationError("StructColumn must have at least one field")
```
Comment (cpcloud) [inline ibis/expr/operations/structs.py:39]: Please remove this change. We haven't decided whether to continue to allow empty structs. For now, they are allowed.

```suggestion
```

File: ibis/expr/operations/structs.py [hunk 23]
```diff
@@ -43,6 +45,5 @@ def __init__(self, names, values):
         super().__init__(names=names, values=values)
 
     @attribute
-    def dtype(self) -> dt.DataType:
-        dtypes = (value.dtype for value in self.values)
-        return dt.Struct.from_tuples(zip(self.names, dtypes))
```
Comment (cpcloud) [inline ibis/expr/operations/structs.py:48]: Can you avoid making style changes in this PR?

File: ibis/expr/types/arrays.py [hunk 24]
```diff
@@ -1106,15 +1115,38 @@ def array(values: Iterable[V]) -> ArrayValue:
...
+            )
+
+    if values is None:
+        if type is None:
+            raise ValidationError("If values is None/NULL, dtype must be provided")
+        return ir.null(type)
+
+    values = tuple(values)
+    if len(values) > 0 and type is not None:
+        return ops.Array(values).to_expr().cast(type)
```
Comment (cpcloud) [inline ibis/expr/types/arrays.py:1150]: Can we avoid doing operations like `cast` here? Doing this makes assumptions about how conversion should work in the backend and casting may or may not be the right approach. Instead, we should handle this in the backends.

It'll probably be extra code to do that.

File: ibis/expr/operations/arrays.py [hunk 25]
```diff
@@ -7,23 +7,32 @@
...
+
+    def __init__(self, exprs, dtype: dt.Array | None = None):
+        if len(exprs) == 0:
+            if dtype is None:
+                raise ValidationError("If values is empty, dtype must be provided")
+            if not isinstance(dtype, dt.Array):
+                raise ValidationError(f"dtype must be an array, got {dtype}")
+        elif dtype is None:
+            dtype = dt.Array(rlz.highest_precedence_dtype(exprs))
+        super().__init__(exprs=exprs, dtype=dtype)
```
Comment (cpcloud) [inline ibis/expr/operations/arrays.py:28]: Why are we doing this validation here and not at the API entry point?

File: ibis/tests/expr/test_literal.py [hunk 26]
```diff
@@ -109,42 +109,29 @@ def test_normalized_underlying_value(userinput, literal_type, expected_type):
...
     assert a.type() == dt.dtype(typestr)
 
 
 @pytest.mark.parametrize(
     "value",
     [
-        dict(field1="value1", field3=3.14),  # wrong field name
-        dict(field1="value1"),  # missing field
+        pytest.param(dict(field1="value1", field3=3.14), id="wrong_field"),
+        pytest.param(dict(field1="value1"), id="missing_field"),
```
Comment (cpcloud) [inline ibis/tests/expr/test_literal.py:119]: import param from pytest to avoid some verbosity

File: ibis/expr/types/structs.py [hunk 27]
```diff
@@ -62,26 +66,50 @@ def struct(
...
+
     fields = dict(value)
-    if any(isinstance(value, Value) for value in fields.values()):
-        names = tuple(fields.keys())
-        values = tuple(fields.values())
-        return ops.StructColumn(names=names, values=values).to_expr()
-    else:
-        return literal(collections.OrderedDict(fields), type=type)
+    names = tuple(fields.keys())
+    values = tuple(fields.values())
```
Comment (cpcloud) [inline ibis/expr/types/structs.py:103]: You shouldn't need to convert these to `tuple`

File: ibis/expr/operations/arrays.py [hunk 28]
```diff
@@ -7,23 +7,32 @@
...
 class Array(Value):
     exprs: VarTuple[Value]
+    dtype: Optional[dt.Array] = None
+
+    def __init__(self, exprs, dtype: dt.Array | None = None):
+        if len(exprs) == 0:
+            if dtype is None:
+                raise ValidationError("If values is empty, dtype must be provided")
+            if not isinstance(dtype, dt.Array):
+                raise ValidationError(f"dtype must be an array, got {dtype}")
```
Comment (cpcloud) [inline ibis/expr/operations/arrays.py:25]: You don't need this at all, the declaration of `dtype: Optional[dt.Array]` creates a validation rule.

File: ibis/expr/operations/arrays.py [hunk 29]
```diff
@@ -7,23 +7,32 @@
...
-from ibis.common.annotations import attribute
+from ibis.common.annotations import ValidationError, attribute
 from ibis.common.typing import VarTuple  # noqa: TCH001
 from ibis.expr.operations.core import Unary, Value
 
 
 @public
 class Array(Value):
     exprs: VarTuple[Value]
+    dtype: Optional[dt.Array] = None
```
Comment (cpcloud) [inline ibis/expr/operations/arrays.py:18]: I think I would prefer to split this operation into two to avoid this messy validation.

One op for inferred type and one for specified type. As is this logic seems unnecessarily complicated.

File: ibis/backends/tests/test_struct.py [hunk 30]
```diff
@@ -79,6 +138,9 @@ def test_all_fields(struct, struct_df):
 
 @pytest.mark.notimpl(["postgres", "risingwave"])
 @pytest.mark.parametrize("field", ["a", "b", "c"])
+@pytest.mark.notyet(
+    ["flink"], reason="flink doesn't support creating struct columns from literals"
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/tests/test_struct.py:142]: I think that's right.

File: ibis/backends/polars/compiler.py (hunk 17 above)
Comment (NickCrews) [reply ibis/backends/polars/compiler.py:90]: Need to look at this again, but IIRC this was actually the minimum possible change, something was erroring with NULL duration handling. Will check again

File: ibis/backends/tests/test_array.py (hunk 21 above)
Comment (NickCrews) [reply ibis/backends/tests/test_array.py:87]: The conversion is exactly the part of the API I was trying to test here. I think this conversion is something that this API should explicitly support. What do you think?

Maybe I add a param as

```python
@pytest.mark.param(raw, [[1,2,3], ["1", "2", "3"]])
def test_array_factory_typed(con, raw):
    typed = ibis.array(raw, type="array<string>")
    ...
```

File: ibis/expr/operations/structs.py (hunk 22 above)
Comment (NickCrews) [reply ibis/expr/operations/structs.py:39]: hmm, yeah IDK why I added this, thanks!

File: ibis/expr/types/arrays.py (hunk 24 above)
Comment (NickCrews) [reply ibis/expr/types/arrays.py:1150]: Not sure I totally understand, like ops.Array should always have a dtype present, and the visit_Array() methods should always deal with it?

I wanted to support specifying a dtype in the ibis.array() API. To get there, I thought it would be better to split up the process into two stages: 1. create the value 2. cast to the right type. This leans better on the already existing casting implementations. But are you saying there are cases where this two-step process doesn't work, ie you lose some precision or something?

File: ibis/expr/operations/arrays.py (hunk 25 above)
Comment (NickCrews) [reply ibis/expr/operations/arrays.py:28]: Currently, ibis.array() is the only entry point for creating an ops.Array. But, I wanted to future proof us: if in the future someone creates an ops.Array from some other place, I wanted them to be able to take advantage of this validation. What do you think? premature optimization?

File: ibis/expr/operations/arrays.py (hunk 28 above)
Comment (NickCrews) [reply ibis/expr/operations/arrays.py:25]: You are right it has a validation rule, but the generic error message isn't that clear.

Also just to be clear (don't think it matters here) the validation from the signature happens during the `super().__init__()` call, so the validation will happen, but in the future, it hasn't happened yet.

File: ibis/expr/operations/arrays.py (hunk 29 above)
Comment (NickCrews) [reply ibis/expr/operations/arrays.py:18]: If the complex validation stuff gets hoisted into ibis.array() (see below), do you still have this concern?

**Comment by NickCrews** [comment]
I'm picking off some of the changes in individual PRs, I'll let you know when this is ready for another look here.

File: ibis/backends/sql/datatypes.py (hunk 20 above)
Comment (NickCrews) [reply ibis/backends/sql/datatypes.py:1024]: done in https://github.com/ibis-project/ibis/pull/9305

File: ibis/expr/operations/structs.py (hunk 22 above)
Comment (NickCrews) [reply ibis/expr/operations/structs.py]: Actually, empty structs are currently disallowed. But working on fixing that in https://github.com/ibis-project/ibis/pull/9310

File: ibis/backends/polars/compiler.py (hunk 17 above)
Comment (NickCrews) [reply ibis/backends/polars/compiler.py:90]: Yes, this reorg is needed. ibis's Duration.DAY and friends are not representable in polars' datatypes, so if you call `PolarsType.from_ibis(dtype)` on those, you get an error. So we need an early guard to look for the interval dtypes and return early.

File: ibis/backends/sql/compiler.py [hunk 31]
```diff
@@ -1019,8 +1019,8 @@ def visit_InSubquery(self, op, *, rel, needle):
             query = sg.select(STAR).from_(query)
         return needle.isin(query=query)
 
-    def visit_Array(self, op, *, exprs):
-        return self.f.array(*exprs)
+    def visit_Array(self, op, *, exprs, dtype):
+        return self.cast(self.f.array(*exprs), dtype)
```
Comment (cpcloud) [inline ibis/backends/sql/compiler.py:1023]: This seems extraordinarily heavy-handed and overly broad.

Do we really need to cast the entire on constructing an array, for all SQL backends?

**Comment by cpcloud** [comment]
This PR is still too big. Can you split the array, map and struct implementations into three separate PRs?

**Comment by NickCrews** [comment]
Yeah splitting it probably is good.

- map is easy, no semantic changes going on there
- for struct, I want to add tests for empty structs. Therefore I think we need to decide on and merge my other PR regarding this behavior 
- for arrays, it's gonna be a little hairy too, because I am implementing empty arrays at the same time as adjusting the API. I don't really want to do them separately or it will just be churn. Unless you see a decent way to address those two things separately

**Comment by cpcloud** [comment]
@NickCrews Just checking in here, can you split these up some more?

**Comment by NickCrews** [comment]
OK, I just tried to split the struct and the map stuff into their own PRs, but both of those rely on the new `ibis.array(arg, typ)` API. So we need to get the array stuff in first. I'll split that out.

**Comment by NickCrews** [comment]
split this into
- #9458 
- #9460 
- #9459
